pkg/core/blockchain.go

```diff
@@ -741,37 +862,208 @@ func (bc *Blockchain) removeOldTransfers(index uint32) time.Duration {
 				acc = batchAcc
-			} else if canDrop { // We've seen this account and all entries in this batch are guaranteed to be outdated.
-				removed++
-				return false
+			} else if edgePassed { // We've seen this account and all entries in this batch are guaranteed to be outdated (if removeOld) / to be relevant (if !removeOld).
+				return postEdge(batchAcc, k, v)
 			}
-			// We don't know what's inside, so keep the current
-			// batch anyway, but allow to drop older ones.
-			canDrop = batchTs <= ts
-			kept++
-			return true
+			// We don't know what's inside, so keep (if removeOld) / drop (if !removeOld) the current
+			// batch anyway, and allow to drop (if removeOld) / keep (if !removeOld) older ones.
+			// Additional processing of the edge batch for precise transfers removal will be performed later.
+			edgePassed = batchTs <= ts
+			return preEdge(batchAcc, k, v, edgePassed)
 		})
 		if err != nil {
 			break
 		}
+
+		if !removeOld {
+			puts := make(map[string][]byte)
+			for batchAcc, log := range edgeLogs {
+				info, ok := transferInfoCache[batchAcc]
+				if !ok {
+					info = state.NewTokenTransferInfo()
+				}
+				var (
+					newBatch      *bool
+					nextBatch     *uint32
+					nextTimestamp *uint64
+				)
+				switch p {
+				case byte(storage.STNEP11Transfers):
+					newBatch = &info.NewNEP11Batch
+					nextBatch = &info.NextNEP11Batch
+					nextTimestamp = &info.NextNEP11NewestTimestamp
+				case byte(storage.STNEP17Transfers):
+					newBatch = &info.NewNEP17Batch
+					nextBatch = &info.NextNEP17Batch
+					nextTimestamp = &info.NextNEP17NewestTimestamp
+				}
+				if log.kv == nil || log.kv.Value[0] == 0 {
+					// We've completely removed the edge log batch. Either there are no batches anymore (then need to delete transfer info)
+					// or there is some more (then need to retrieve the most fresh batch and store proper transfer info for it).
+					if log.nextOlderLog == nil {
+						*newBatch = true
+						*nextBatch = 0
+						*nextTimestamp = 0
+						transferInfoCache[batchAcc] = info
+					} else {
+						// TODO: here we need to iterate over the nextOlderLog transfers and save proper nextBatch/nextTimestamp.
+					}
+				} else {
+					// A part of removed edge log batch is still relevant and should be stored again.
+					puts[string(log.kv.Key)] = log.kv.Value
+
+					// Create proper token transfer info.
+					batchTmstm := binary.BigEndian.Uint64(log.kv.Key[1+util.Uint160Size:])
+					batchIdx := binary.BigEndian.Uint32(log.kv.Key[1+util.Uint160Size+8:])
+					if log.kv.Value[0] >= state.TokenTransferBatchSize {
+						*newBatch = true
+						*nextBatch = batchIdx + 1
+						*nextTimestamp = log.newestTransferTimestamp
+					} else {
+						*newBatch = false
+						*nextBatch = batchIdx
+						*nextTimestamp = batchTmstm
+					}
+					transferInfoCache[batchAcc] = info
+				}
+			}
+			bc.store.PutChangeSet(puts, nil)
+		}
 	}
+
+	// Update or remove token transfer info.
```

Comment:
Or maybe we need a completely different strategy here. Use _forward_ seek and rebuild TTI on the fly, then strip newer entries flushing TTI for an account. Currently `removeTransfers` looks very hairy.

Reply:
Implemented.